client/src/api/block_builder/input_selection/new/remainder.rs

```diff
@@ -0,0 +1,111 @@
+    requirement::{
+        base_token::base_token_sums,
+        native_tokens::{get_minted_and_melted_native_tokens, get_native_tokens, get_native_tokens_diff},
+    },
+    OutputInfo,
+};
+use crate::{
+    block::{
+        address::Address,
+        output::{
+            unlock_condition::{AddressUnlockCondition, UnlockCondition},
+            BasicOutputBuilder, Output,
+        },
+        protocol::ProtocolParameters,
+    },
+    error::{Error, Result},
+    secret::types::InputSigningData,
+};
+
+// Gets the remainder address from configuration of finds one from the inputs.
+// TODO should this also look for non-ed25519 addresses?
+// TODO should this also check available inputs ? not only selected.
+fn get_remainder_address(selected_inputs: &[InputSigningData], remainder_address: Option<Address>) -> Option<Address> {
+    if remainder_address.is_some() {
+        return remainder_address;
+    }
+
+    // TODO need to check timelock/expiration?
+
+    for input in selected_inputs {
+        if let Some(unlock_conditions) = input.output.unlock_conditions() {
+            if let Some(address_unlock_condition) = unlock_conditions.address() {
+                if address_unlock_condition.address().is_ed25519() {
+                    return Some(*address_unlock_condition.address());
+                }
+            }
+
+            // TODO check transition type
+
+            if let Some(governor_address_unlock_condition) = unlock_conditions.governor_address() {
+                if governor_address_unlock_condition.address().is_ed25519() {
+                    return Some(*governor_address_unlock_condition.address());
+                }
+            }
+
+            if let Some(state_controller_address_unlock_condition) = unlock_conditions.state_controller_address() {
+                if state_controller_address_unlock_condition.address().is_ed25519() {
+                    return Some(*state_controller_address_unlock_condition.address());
+                }
+            }
+        }
+    }
+
+    None
+}
+
+pub(crate) fn remainder_output(
+    selected_inputs: &[InputSigningData],
+    outputs: &[OutputInfo],
+    remainder_address: Option<Address>,
+    protocol_parameters: &ProtocolParameters,
+) -> Result<Option<Output>> {
+    let (inputs_sum, outputs_sum) = base_token_sums(selected_inputs, outputs);
+
+    let mut input_native_tokens = get_native_tokens(selected_inputs.iter().map(|input| &input.output))?;
+    let mut output_native_tokens = get_native_tokens(outputs.iter().map(|output| &output.output))?;
+    let (minted_native_tokens, melted_native_tokens) = get_minted_and_melted_native_tokens(selected_inputs, outputs)?;
+
+    input_native_tokens.merge(minted_native_tokens)?;
+    output_native_tokens.merge(melted_native_tokens)?;
+    // TODO also merge burn
+
+    let native_tokens_diff = get_native_tokens_diff(&input_native_tokens, &output_native_tokens)?;
+
+    if inputs_sum > outputs_sum || native_tokens_diff.is_some() {
```

Comment:
`if native_tokens_diff.is_some()`, but `inputs_sum <= outputs_sum` we would need to check if we can add more inputs, which is currently not happening, right?

Reply:
It is not possible at this point (in the remainder function), that `inputs_sum < outputs_sum`, and if it is equal then there is nothing to be done